trigger.tex

```diff
@@ -142,33 +142,84 @@ \section{Priority}
-that interrupts might be disabled at other times when a user might want
-triggers to fire.
-
-A more complex solution is to implement \FcsrTcontrolMte and \FcsrTcontrolMpte in \RcsrTcontrol. This
-solution has the benefit that it only disables triggers during the trap
-handler.
-
-A user setting M-mode triggers that cause breakpoint exceptions will have to be
-aware of any problems that might come up with the particular hart they are
-working on.
+\section{Native Triggers}
+\label{sec:nativetrigger}
+
+\begin{commentary}
+Triggers can be used for native debugging when \FcsrMcontrolSixAction=0.
+If supported by the hart and desired by the debugger, triggers will often be
+programmed to have \FcsrMcontrolSixM=0 so that when they fire they cause a
+breakpoint exception to trap to a more privileged mode. That breakpoint
+exception can either be taken in M-mode or it can be delegated to a less
+privileged mode. However, it is possible for triggers to fire in the same
+mode that the resulting exception will be handled in.
+
+\begin{steps}{In particular, when \FcsrMcontrolSixAction=0:}
+\item mcontrol and mcontrol6 triggers with \FcsrMcontrolSixM=1 can cause a
+breakpoint exception that is taken from M-mode to M-mode (regardless of
+delegation).
+\item mcontrol and mcontrol6 triggers with \FcsrMcontrolSixS=1 can cause a
+breakpoint exception that is taken from S-mode to S-mode if \Rmedeleg[3]=1.
+\item mcontrol6 triggers with \FcsrMcontrolSixVs=1 can cause a
+breakpoint exception that is taken from VS-mode to VS-mode if \Rmedeleg[3]=1
+and \Rhedeleg[3]=1.
+\item icount triggers with \FcsrIcountM=1 can cause a
+breakpoint exception that is taken from M-mode to M-mode (regardless of
+delegation).
+\item icount triggers with \FcsrIcountS=1 can cause a
+breakpoint exception that is taken from S-mode to S-mode if \Rmedeleg[3]=1.
+\item icount triggers with \FcsrIcountVs=1 can cause a
+breakpoint exception that is taken from VS-mode to VS-mode if \Rmedeleg[3]=1
+and \Rhedeleg[3]=1.
+\item etrigger and itrigger triggers will always be taken from a trap handler
+before the first instruction of the handler.  If etrigger/itrigger is set to
+trigger on exception/interrupt X and if X is delegated to mode Y then the
+trigger will cause a breakpoint exception that is taken from mode Y to mode
+Y unless breakpoint exceptions are delegated to a more privileged mode than Y.
+\item tmexttrigger triggers are asynchronous and may occur in any mode and
+at any time.
+\end{steps}
+
+In these cases such a trigger may cause a breakpoint exception while already
+in a trap handler. This might leave the hart unable to resume normal execution
+because state such as \Rmcause and \Rmepc would be overwritten.
+\end{commentary}
+
+\begin{steps}{Harts that support triggers with \FcsrMcontrolSixAction=0
+should implement one of the following two solutions to solve the problem of
+reentrancy:}
+\item The hardware must prevent triggers with \FcsrMcontrolSixAction=0 from
```

Comment:
Must in these two items is confusing. I'd leave the line above (which now says should) and then describe the behavior without must/should. E.g. "The hardware prevents triggers ..."

Reply:
I obviously didn't reread the whole thing when making that must->should change.  I'll reread the whole thing and make changes along the lines of your suggestion.